torch/_subclasses/functional_tensor.py

```diff
@@ -0,0 +1,259 @@
+        # Indicates to our torch_dispatch dispatching infra that
+        # this is an "infra" mode with lower dispatching precedence.
+        self._mode_key = torch._C._TorchDispatchModeKey.FUNCTIONAL
+
+    # No-op if FunctionalTensorMode is already in use
+    def __enter__(self):
+        if (
+            torch._C._get_dispatch_mode(torch._C._TorchDispatchModeKey.FUNCTIONAL)
+            is None
+        ):
+            self.enter_stack.append(True)
+            return super().__enter__()
+        else:
+            self.enter_stack.append(False)
+            return self
+
+    def __exit__(self, a, b, c):
+        is_on_stack = self.enter_stack.pop()
+        if is_on_stack:
+            super().__exit__(a, b, c)
+
+    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+
+        unrecognized_types = [
+            t
+            for t in types
+            if not issubclass(t, torch._subclasses.FakeTensor)
+            and t not in [torch.Tensor, FunctionalTensor]
+        ]
+        if unrecognized_types:
+            not_implemented_log.debug(
+                "FunctionalTensor unrecognized subclass(es): %s", unrecognized_types
+            )
+            return NotImplemented
+
+        def assert_is_functional(x):
+            assert torch._is_functional_tensor(x)
+
+        def wrap(x):
+            return FunctionalTensor(x)
+
+        any_functional_inputs = False
+
+        def unwrap(x):
+            any_functional_inputs = True
+            return x.elem
+
+        args_unwrapped, kwargs_unwrapped = pytree.tree_map_only(
+            FunctionalTensor, unwrap, (args, kwargs)
+        )
+
+        if not self.is_active:
+            assert not any_functional_inputs
+            return func(*args_unwrapped, **kwargs)
+
+        # Expectation: functionalization should not **already** be enabled above our mode.
+        # Why would that be bad? when we return a FunctionalTensor here, we don't want functionalization
+        # to run above this mode and further wrap that output in **another** C++ FunctionalTensorWrapper.
+        is_included = torch._C._dispatch_tls_is_dispatch_key_included(
+            torch._C.DispatchKey.Functionalize
+        )
+        is_excluded = torch._C._dispatch_tls_is_dispatch_key_excluded(
+            torch._C.DispatchKey.Functionalize
+        )
+        assert is_excluded or not is_included
+        # All we want to do here is re-use the existing C++ functionalization logic.
+        # This requires swizzling our TLS dispatch keys so that the Functionalize key is active.
+        with torch._C._SetExcludeDispatchKeyGuard(
+            torch._C.DispatchKey.Functionalize, False
+        ), torch._C._IncludeDispatchKeyGuard(torch._C.DispatchKey.Functionalize):
+            try:
+                # By default for python functionalization (for AOTAutograd), we reapply views.
+                torch._functionalize_enable_reapply_views(True)
```

Comment:
This should be scoped right? You need to reset it to its old value when you exit this.